### PR TITLE
[ES|QL] Add CombineBinaryComparisons and Range

### LINKED ISSUE #108525: [CI] LogicalPlanOptimizerTests testSimplifyComparisonArithmeticWithDisjunction failing
I haven't investigated this one yet.  There are several other issues around how this optimization works, and it's possible fixing one of them will fix this test.

**Build scan:**
https://gradle-enterprise.elastic.co/s/zn7kv6lwmdrow/tests/:x-pack:plugin:esql:test/org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests/testSimplifyComparisonArithmeticWithDisjunction

**Reproduction line:**
```
./gradlew ':x-pack:plugin:esql:test' --tests "org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.testSimplifyComparisonArithmeticWithDisjunction" -Dtests.seed=A8AF70E4F3429963 -Dtests.locale=zh-TW -Dtests.timezone=Asia/Chungking -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests#testSimplifyComparisonArithmeticWithDisjunction`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testSimplifyComparisonArithmeticWithDisjunction'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.lang.AssertionError: Expected filter condition to be a binary comparison but found [12 * (-integer - 5) >= -120 OR integer < 5]

  at __randomizedtesting.SeedInfo.seed([A8AF70E4F3429963:7A91B4839C45DBAE]:0)
  at org.junit.Assert.fail(Assert.java:89)
  at org.junit.Assert.assertTrue(Assert.java:42)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.extractPlannedBinaryComparison(LogicalPlanOptimizerTests.java:4405)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.extractPlannedBinaryComparison(LogicalPlanOptimizerTests.java:4397)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.doTestSimplifyComparisonArithmetics(LogicalPlanOptimizerTests.java:4418)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.testSimplifyComparisonArithmeticWithDisjunction(LogicalPlanOptimizerTests.java:4512)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by fang-xing-esql**
This is another option to resolve #108525, a more complete one comparing to #110548.

As discussed in #110548 , the complete CombineBinaryComparisons rule creates Range(#110548 removed it), which is not a supported operator in ES|QL yet. Range can be pushed down into ES query, although it may not always applicable. If adding the the complete CombineBinaryComparisons rule into ES|QL, we might want a full implementation of Range in ES|QL as well, which is addressed in this PR.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/Range.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            in.readNamedWriteable(Expression.class),
+            in.readBoolean(),
+            in.readNamedWriteable(Expression.class),
+            in.readBoolean(),
+            in.readZoneId()
+        );
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/Range.java:72]: if (out.getVersion().onOrAfteR(kjbdafdasf) {
  new And(new GreaterThan())))).writeTo(out);
  return;
}

**Review by astefan — COMMENTED** [review]
Adding the Range might make it easier for us to reason about two comparisons applied to the same attribute (a > 5 and a < 10) and it also helps pushing down to Lucene a more appropriate `range` query where both `lt/lte` and `gt/gte` are used.

But this comes with the drawback of leaking a pushdown (physical plan) optimization into the logical plan even though the intention of the query is different (there is no element in the query that is parsed as a `Range` logical plan).

Until we add a syntactic sugar like `field between 1 and 10` or maybe `1 < field < 10` I suggest not using the `Range` in the logical and physical plans and consider a better `range` pushdown optimization as a follow up issue.

**Comment by fang-xing-esql** [comment]
> Adding the Range might make it easier for us to reason about two comparisons applied to the same attribute (a > 5 and a < 10) and it also helps pushing down to Lucene a more appropriate `range` query where both `lt/lte` and `gt/gte` are used.
> 
> But this comes with the drawback of leaking a pushdown (physical plan) optimization into the logical plan even though the intention of the query is different (there is no element in the query that is parsed as a `Range` logical plan).
> 
> Until we add a syntactic sugar like `field between 1 and 10` or maybe `1 < field < 10` I suggest not using the `Range` in the logical and physical plans and consider a better `range` pushdown optimization as a follow up issue.

Thanks for reviewing @astefan ! There are another two PRs to reduce the scope, I just saw @bpintea approved #110548 too before making #111437 ready for review. I'll merge #110548, if #111437 looks good I'll merge it as a follow up PR.

#110548 Add `CombineBinaryComparisons` without `Range` 
#111437 Add `CombineBinaryComparisons` without `Range` in `LocalPhysicalPlan` and `ExpressionTranslators`

**Comment by astefan** [comment]
Since https://github.com/elastic/elasticsearch/pull/110548 has been merged, I'll go ahead and do a bit of cleanup and close this PR.

**Pull request closed without merging**